File: python/백준/다이나믹 프로그래밍/1535 안녕.py
import sys
input = sys.stdin.readline

N = int(input())
L = list(map(int, input().split()))  # 잃는 체력
J = list(map(int, input().split()))  # 얻는 기쁨

MAX_LOSS = 99  # 총 잃을 수 있는 체력의 최대값

dp = [0] * (MAX_LOSS + 1)  # dp[h] = 체력을 h만큼 잃었을 때의 최대 기쁨

for i in range(N):
    lose = L[i]
    joy = J[i]
    # 0/1 배낭: 내림차순으로 순회
    for h in range(MAX_LOSS, lose - 1, -1):
        dp[h] = max(dp[h], dp[h - lose] + joy)

print(max(dp))  

File: cpp/백준/다이나믹 프로그래밍/1535 안녕.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    int N;
    cin >> N;

    vector<int> L(N), J(N);
    for (int i = 0; i < N; ++i) cin >> L[i];
    for (int i = 0; i < N; ++i) cin >> J[i];

    const int MAX_LOSS = 99;
    vector<int> dp(MAX_LOSS + 1, 0);

    for (int i = 0; i < N; ++i) {
        int lose = L[i];
        int joy = J[i];
        for (int h = MAX_LOSS; h >= lose; --h) {
            dp[h] = max(dp[h], dp[h - lose] + joy);
        }
    }

    cout << *max_element(dp.begin(), dp.end()) << "\n";
    return 0;
}
